src/wallet/wallet.ts

```diff
@@ -103,170 +142,442 @@ class HathorWalletServiceWallet extends EventEmitter {
         } else if (data.status.status === 'ready') {
-          this.setState(walletState.READY);
+          await this.onWalletReady();
         } else {
-          // TODO What other status might have?
-          // throw error?
+          throw new WalletRequestError('Error getting wallet status.');
         }
       } else {
-        // TODO is there any other possible error to handle here?
+        throw new WalletRequestError('Error getting wallet status.');
       }
     } catch(err) {
-      // TODO How to handle error
-      console.log('Error sending create wallet request', err);
+      // Should we handle any error?
+      throw err;
+    }
+  }
+
+  /**
+   * Check if wallet is ready and throw error if not ready
+   *
+   * @memberof HathorWalletServiceWallet
+   * @inner
+   */
+  private checkWalletReady() {
+    if (!this.isReady()) {
+      throw new WalletError('Wallet not ready');
     }
   }
 
+  /**
+   * Method executed when wallet is ready
+   *
+   * @memberof HathorWalletServiceWallet
+   * @inner
+   */
+  private async onWalletReady() {
+    this.setState(walletState.READY);
+    await this.getNewAddresses();
+  }
+
   /**
    * Get all addresses of the wallet
    *
-   * @memberof HathorWallet
+   * @memberof HathorWalletServiceWallet
    * @inner
    */
-  async getAllAddresses(): Promise<string[]> {
-    const response = await walletApi.getAddresses(this.walletId!);
-    let addresses = [];
+  async getAllAddresses(): Promise<GetAddressesObject[]> {
+    this.checkWalletReady();
+    const response = await walletApi.getAddresses(this);
+    let addresses: GetAddressesObject[] = [];
     if (response.status === 200 && response.data.success === true) {
       addresses = response.data.addresses;
     } else {
-      // TODO What error should be handled here?
+      throw new WalletRequestError('Error getting wallet addresses.');
     }
     return addresses;
   }
 
+  private async getNewAddresses() {
+    this.checkWalletReady();
+    const response = await walletApi.getNewAddresses(this);
+    let addresses: AddressInfoObject[] = [];
+    if (response.status === 200 && response.data.success === true) {
+      addresses = response.data.addresses;
+      this.newAddresses = addresses;
+      this.indexToUse = 0;
+    } else {
+      throw new WalletRequestError('Error getting wallet addresses to use.');
+    }
+  }
+
   /**
    * Get the balance of the wallet for a specific token
    *
-   * @memberof HathorWallet
+   * @memberof HathorWalletServiceWallet
    * @inner
    */
-  async getBalance(token: string | null = null) {
-    const response = await walletApi.getBalances(this.walletId!, token);
-    let balance = null;
+  async getBalance(token: string | null = null): Promise<GetBalanceObject[]> {
+    this.checkWalletReady();
+    const response = await walletApi.getBalances(this, token);
+    let balance: GetBalanceObject[] = [];
     if (response.status === 200 && response.data.success === true) {
```

Comment:
Shouldn't we move this logic to the walletApi? I think this code must be way simpler to read because it would have to deal with a promise that might succeed or fail.

Reply:
As far as I checked, you've already refactored it as well, right?

Reply:
Yes! I forgot to reply this one, sorry